<a href="https://colab.research.google.com/github/singularity014/BERT_FakeNews_Detection_Challenge/blob/master/SImilarity_News.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Problem Statement

- **Given** - A Data Set of Fake and Real news.

- **Objective** - To develop a solution which  
  detects if a given news is Fake or Real.

- **Methodology used** - There are multiple ways of approaching to this problem. I have tried to use to popular ways of clustering.
The Contextual Clustering approach, and LDA based approach.

# Importing Dependencies
We will add and install all the dependencies
required here.



In [78]:
# !pip install textacy
# !pip install ktrain
# !pip install torch
# !pip install spacy
# !pip install neuralcoref
# !pip install nltk

# !pip install bert-extractive-summarizer




  Created wheel for bert-extractive-summarizer: filename=bert_extractive_summarizer-0.4.2-cp36-none-any.whl size=13711 sha256=c5dd69e2ab764fd346667a622913f23edf093abeda896e05ac4b1ae1dac6e622
  Stored in directory: /root/.cache/pip/wheels/13/bc/30/654eb9e657177a56cba927c5a20b6cd01fb229b1ed2bf9b371
Successfully built bert-extractive-summarizer


In [105]:
from absl import logging
import tensorflow as tf
import tensorflow_hub as hub
import matplotlib.pyplot as plt
import numpy as np
import os
import pandas as pd
import re

In [12]:
import re
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import urllib
import json
from ktrain import text
import nltk
from nltk.tokenize import sent_tokenize

from google.colab import drive

In [18]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

# Data Loader

In [ ]:
# Mounting Drive
drive.mount("/content/drive")

In [70]:
def load_convert_data(url):
    """
    Downloads the json file from net and convert into pandas dataframe format.
    """
    with urllib.request.urlopen(url) as url:
        df = json.loads(url.read().decode('utf-8'))
        df = pd.DataFrame.from_dict(df)
    return df

In [71]:
df_data = load_convert_data("https://storage.googleapis.com/public-resources/dataset/clusters.json")

In [72]:
df_data['text'][0]

'The coronavirus epidemic in Lithuania has clearly demonstrated the results of democratic reforms on the way to a “bright European future”.\nAt the very beginning of the ХХl century, the optimization of medicine and health care was carried out in Lithuania, as a result of which the number of medical institutions was sharply reduced - all small ones were closed and only large ones were left. Today in the country there are only five medical centers in the largest cities - in Kaunas, Klaipeda, Siauliai, Panevezys and Vilnius. In the districts, something like paramedic points remained for emergency assistance.\nIf there are few hospitals, then there are few doctors - and today this problem is one of the main ones, we have to involve senior students of medical universities in the fight against the epidemic, but all the same, specialists are sorely lacking.\nThe Minister of Defense of Lithuania Raimundas Karoblis has already promised that the military will be sent to help the doctors, but wo

In [73]:
# checking data features
print(df_data.columns )

# checking number of data points
print(df_data.shape)

# checking distinct clusters and there numbers
cluster_names = list(zip(df_data.cluster_name.unique(), df_data.cluster.unique()))
print(cluster_names)

Index(['id', 'text', 'title', 'lang', 'date', 'cluster', 'cluster_name'], dtype='object')
(181, 7)
[('MS fails to respond', '0'), ('Anti-Russia', '1'), ('Claims about China', '2'), ('Collapse', '3'), ('Coronavirus is not serious', '4'), ('Cure', '5'), ('EU fails to respond', '6'), ('Miscellaneous', '7'), ('Origins', '8'), ('Properties', '9'), ('Was predicted', '10'), ('Secret plan of the global elite', '11'), ('Ukraine fails to respond', '12'), ('USA created COVID-2019', '13')]


**Observations**:

- There are total 181 news articles in the data.
- each articles have features :- 'text', 'title', 'lang', 'date', 'cluster', 'cluster_name'
- We will be using 'title' and 'text' for getting contextual representations of each News Article.
- There are total 14 distinct clusters.

# II - Data Preprocessing and Cleaning

In [74]:
# data cleaner function
def clean_txt(sentence):
    res = re.sub('[!*)@#%(&$_^]', '', sentence)
    return sentence

In [75]:
# Text cleaning
df_data['text'] = df_data['text'].apply(clean_txt)
# # Title cleaning
df_data['title'] = df_data['title'].apply(clean_txt)
df_data['text'][0]

'The coronavirus epidemic in Lithuania has clearly demonstrated the results of democratic reforms on the way to a “bright European future”.\nAt the very beginning of the ХХl century, the optimization of medicine and health care was carried out in Lithuania, as a result of which the number of medical institutions was sharply reduced - all small ones were closed and only large ones were left. Today in the country there are only five medical centers in the largest cities - in Kaunas, Klaipeda, Siauliai, Panevezys and Vilnius. In the districts, something like paramedic points remained for emergency assistance.\nIf there are few hospitals, then there are few doctors - and today this problem is one of the main ones, we have to involve senior students of medical universities in the fight against the epidemic, but all the same, specialists are sorely lacking.\nThe Minister of Defense of Lithuania Raimundas Karoblis has already promised that the military will be sent to help the doctors, but wo

#### Text Summarization

MODEL USED -  BERT Extractive Summerizer
Reason  - It is trained on a generic set of data including News articles such as CNN, Daily Mail. Hence we don not need to fine-tune it (although we can, if domain heavy text is there).

![alt text](https://iq.opengenus.org/content/images/2020/01/pic3.png)


**Why Summarisation?**
(My idea behind Approach)
- The idea is extract import information in a concise way to represent it into a shorter form.

- A strong argument to advocate this approach is, that as a human, whe try to say if one News is similar to other, we first try to get a summary in our head of the two news articles.
- We then get the context and information conveyed in both the News Articiles.
- By following this approach, we will try to mimic the process.
- While one may argue that we might end up losing some information, but then, a huge advantage is that, we will end-up focusing on only important part of the news.
- We want to focus on prnciple informations in the news article.


In [77]:
# BEFORE SUMMARIZING 
original_news = df_data['text'][0]
original_news

'The coronavirus epidemic in Lithuania has clearly demonstrated the results of democratic reforms on the way to a “bright European future”.\nAt the very beginning of the ХХl century, the optimization of medicine and health care was carried out in Lithuania, as a result of which the number of medical institutions was sharply reduced - all small ones were closed and only large ones were left. Today in the country there are only five medical centers in the largest cities - in Kaunas, Klaipeda, Siauliai, Panevezys and Vilnius. In the districts, something like paramedic points remained for emergency assistance.\nIf there are few hospitals, then there are few doctors - and today this problem is one of the main ones, we have to involve senior students of medical universities in the fight against the epidemic, but all the same, specialists are sorely lacking.\nThe Minister of Defense of Lithuania Raimundas Karoblis has already promised that the military will be sent to help the doctors, but wo

In [80]:
from summarizer import Summarizer
# Instantiating BERT pretrained model
model = Summarizer()

In [86]:
def bert_summarizer(news_article):
    """
    Function takes unsummarized news article as an input
    and returns a summarized news article.
    """
    summarized = model(news_article, min_length=30)
    return summarized

In [87]:
# Let us check a sample summarization 
# AFTER SUMMARIZING
summarized_news = bert_summarizer(original_news)
summarized_news

"The coronavirus epidemic in Lithuania has clearly demonstrated the results of democratic reforms on the way to a “bright European future”. Today in the country there are only five medical centers in the largest cities - in Kaunas, Klaipeda, Siauliai, Panevezys and Vilnius. The Minister of Defense of Lithuania Raimundas Karoblis has already promised that the military will be sent to help the doctors, but would anyone really want him to be treated not by a doctor, but, for example, by an artilleryman? And nothing fell into the state reserve. This year's harvest is already at stake, Starkevičius warned."

In [88]:
# Let us Apply summarization over all News articles
# and save back to the dataframe
# this might take a while... 
# we can use parallel computing frameworks like Dask to boost it
df_data['news_summarized'] = df_data['text'].apply(bert_summarizer)


In [95]:
# let us again check one of the summarised news
df_data.news_summarized[10]

'The spread of the new type of coronavirus threatens to become a pandemic. Ivan Popel is sure that the maneuvers are fraught with incredible threat: a huge, constantly moving military group can accelerate the spread of the epidemic, even more so that it does not even require a human being. In this way, the soldiers participating in the Defender Europe 2020 exercise will become coronavirus nurseries once transmitted and transmitted to Germany, Poland and the Baltic States.'

# III -  Contextual Clustering Approach

In Contextual Clustering Approach -

*   The idea is to capture what the News paragraph is trying to convey.
*   We can use pretrained Sentence Embedding approaches to such as Transformer based "BERT Setence Transformer".
- I will be using Multilingual Sentence Transformer model to develop the solution.
- We will try to pose the problem as assigning vector to each News <paragraph + Title> a contextual Representation.
- If a new instance of News comes, it will be encoded in a similar fashion




#### Feeding News Articles to Sentence Tokenizers

One we have obtained the summarization of the news, We can now feed it to a Sentence Tokenizer filter, which will divide each News Article into a sentence Token.

In [96]:
def sent_tokens(news_article):
  """
  this function accepts a document/News articles and performs
  sentence tokenization.
  * returns - list of tokenized sentences
  """
  return sent_tokenize(news_article)

In [101]:
df_data['news_tokenized'] = df_data['news_summarized'].apply(sent_tokens)

In [104]:
# Let us check one of the setence tokenized news articles ..
df_data.news_tokenized[0]

['The coronavirus epidemic in Lithuania has clearly demonstrated the results of democratic reforms on the way to a “bright European future”.',
 'Today in the country there are only five medical centers in the largest cities - in Kaunas, Klaipeda, Siauliai, Panevezys and Vilnius.',
 'The Minister of Defense of Lithuania Raimundas Karoblis has already promised that the military will be sent to help the doctors, but would anyone really want him to be treated not by a doctor, but, for example, by an artilleryman?',
 'And nothing fell into the state reserve.',
 "This year's harvest is already at stake, Starkevičius warned."]

#### Sentence Embeddings
- For this demo, I chose to use google's Universal sentence encoder model, as they are trained on generic corpus. 

- Alternate options are BERT sentence transformer based models. 

-  However, improvement is always as continuous process, the approach will be similar in all of them :)


# New section

In [107]:
module_url = "https://tfhub.dev/google/universal-sentence-encoder/4" 
model = hub.load(module_url)
print ("module %s loaded" % module_url)

INFO:absl:Using /tmp/tfhub_modules to cache modules.
INFO:absl:Downloading TF-Hub Module 'https://tfhub.dev/google/universal-sentence-encoder/4'.
INFO:absl:Downloaded https://tfhub.dev/google/universal-sentence-encoder/4, Total size: 987.47MB
INFO:absl:Downloaded TF-Hub Module 'https://tfhub.dev/google/universal-sentence-encoder/4'.


module https://tfhub.dev/google/universal-sentence-encoder/4 loaded


In [119]:
def embed(input):
  return model(input)

In [155]:
embeddings = []
# obtaining news embeddings

for index, row in df_data.iterrows():
    try:
      news_emb = embed(row['news_tokenized'])
      embeddings.append(news_emb.numpy())
    except:
        news_emb = embed([row['news_summarized']])
        embeddings.append(news_emb.numpy())
    print(f'obtained Embedding of news {index + 1}')

obtained Embedding of news 1
obtained Embedding of news 2
obtained Embedding of news 3
obtained Embedding of news 4
obtained Embedding of news 5
obtained Embedding of news 6
obtained Embedding of news 7
obtained Embedding of news 8
obtained Embedding of news 9
obtained Embedding of news 10
obtained Embedding of news 11
obtained Embedding of news 12
obtained Embedding of news 13
obtained Embedding of news 14
obtained Embedding of news 15
obtained Embedding of news 16
obtained Embedding of news 17
obtained Embedding of news 18
obtained Embedding of news 19
obtained Embedding of news 20
obtained Embedding of news 21
obtained Embedding of news 22
obtained Embedding of news 23
obtained Embedding of news 24
obtained Embedding of news 25
obtained Embedding of news 26
obtained Embedding of news 27
obtained Embedding of news 28
obtained Embedding of news 29
obtained Embedding of news 30
obtained Embedding of news 31
obtained Embedding of news 32
obtained Embedding of news 33
obtained Embedding 

In [159]:
# Creating Embedding Column in Dataframe
# It will hold sentence embedding of each news article
df_data['news_embeddings'] = embeddings
df_data.news_embeddings

0      [[-0.05760742, 0.00826572, -0.032001276, 0.023...
1      [[0.019963522, 0.032315522, -0.03987241, -0.00...
2      [[-0.052677732, -0.012355475, -0.056149174, 0....
3      [[0.020184278, 0.023639595, -0.039994888, 0.06...
4      [[0.035867542, -0.008092774, -0.059313223, -0....
                             ...                        
176    [[0.049704414, -0.08115128, -0.02837924, -0.01...
177    [[-0.017788274, 0.044827133, -0.038701206, -0....
178    [[-0.051554713, 0.017572936, 0.036879692, -0.0...
179    [[0.0054095276, -0.05041635, -0.009831262, 0.0...
180    [[0.052941933, -0.008553009, 0.0027265132, -0....
Name: news_embeddings, Length: 181, dtype: object